File: today/today.py
import sys

from rich.console import Console

from today.cli import build_parser, parse_args, parse_task_files, maybe_display_specific_task, tasks_to_tree


def run(args) -> None:
    parser = build_parser()
    cli_args = parse_args(parser, args)
    console = Console()

    tasks = parse_task_files(cli_args)
    maybe_display_specific_task(cli_args, tasks, console)  # If a specific task is displayed, the program will exit

    tree = tasks_to_tree(cli_args, tasks)
    console.print("")
    console.print(tree)
    console.print("")


def main():
    sys.exit(run(sys.argv[1:]))

File: src/bin/today.rs
use clap::Parser;
use today::cli::{CliArgs, parse_task_files, display_specific_task, display_tasks_tree};
use std::process;

fn main() {
    if let Err(e) = run() {
        eprintln!("Error: {}", e);
        process::exit(1);
    }
}

fn run() -> Result<(), String> {
    let args = CliArgs::parse();
    let today = args.get_today()?;

    // Parse all task files
    let tasks = parse_task_files(&args)?;

    // If a specific task_id is provided, display that task
    if let Some(task_id_result) = args.get_task_id() {
        match task_id_result {
            Ok(task_id) => {
                if task_id >= tasks.len() {
                    return Err(format!("Task ID {} does not exist (only {} tasks found)", task_id, tasks.len()));
                }
                display_specific_task(&tasks[task_id], today);
            }
            Err(task_str) => {
                return Err(format!("Invalid task ID: {}", task_str));
            }
        }
    } else {
        // Display all tasks as a tree
        display_tasks_tree(&args, &tasks)?;
    }

    Ok(())
}
